### ISSUE TITLE
Regression: `operator does not exist: character varying[] = text[]` when querying varchar array column by value

### ISSUE THREAD

**Issue by papandreou** [sequelize/sequelize#16391]
## Issue Creation Checklist

- [x] I understand that my issue will be automatically closed if I don't fill in the requested information
- [x] I have read the [contribution guidelines](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md)

## Bug Description

Querying a `VARCHAR(...)[]` column by a literal array value using `Model#find...` stopped working, it generates a query that errors with `operator does not exist: character varying[] = text[]`.

`git bisect` says that it broke in #15598, released in [v7.0.0-alpha.24](https://github.com/sequelize/sequelize/releases/tag/v7.0.0-alpha.24):

```
50898cac7c979edd94cd7eb68242d9aff7362378 is the first bad commit
commit 50898cac7c979edd94cd7eb68242d9aff7362378
Author: Zoé <zoe@ephys.dev>
Date:   Sun Mar 12 11:09:03 2023 +0100

    feat: rewrite the part of QueryGenerator responsible for WHERE (#15598)
```

Before it would generate a query like this:

```sql
SELECT ... FROM "my_models" AS "MyModel" WHERE "MyModel"."scopes" = ARRAY['foo','bar']::VARCHAR(64)[] LIMIT 1;
```

After #15598 was merged, it comes out as:

```sql
SELECT ... FROM "my_models" AS "MyModel" WHERE "MyModel"."scopes" = ARRAY['foo','bar'] LIMIT 1;
```

... which apparently causes Postgres to not be able to match up the types. I have not tested this with any other dialects.

### Reproducible Example

```js
import Sequelize from '@sequelize/core';

const sequelize = new Sequelize({
  dialect: 'postgres',
  // ... credentials
});

class MyModel extends Sequelize.Model {}

MyModel.init(
  {
    myModelId: {
      type: Sequelize.DataTypes.BIGINT,
      primaryKey: true,
    },
    scopes: {
      type: Sequelize.DataTypes.ARRAY(Sequelize.DataTypes.STRING(64)),
    },
  },
  {
    sequelize,
    tableName: 'my_models',
    timestamps: true,
  },
);

await sequelize.sync();

await sequelize.models.MyModel.findOne({
  where: {
    scopes: ['foo', 'bar'],
  },
});
```

### What do you expect to happen?

That Sequelize runs a valid and correct query that does not result in an error.

### What is actually happening?

Sequelize runs this query:

```sql
SELECT "myModelId", "scopes", "createdAt", "updatedAt" FROM "my_models" AS "MyModel" WHERE "MyModel"."scopes" = ARRAY['foo','bar'] LIMIT 1;
```

which results in this error (code: `42883`):

```
operator does not exist: character varying[] = text[]
No operator matches the given name and argument types. You might need to add explicit type casts.
```

### Environment

- Sequelize version: 7.0.0-alpha.27, but also reproducible with 4d624fd37e84a0b9215bac0b0a1ab85892a81cef, which is the current tip of `main`
- Node.js version: 18.16.1
- If TypeScript related: TypeScript version: 4.9.5
- Database & Version: PostgreSQL 15.3
- Connector library & Version: `pg`

## Would you be willing to resolve this issue by submitting a Pull Request?

- [ ] Yes, I have the time and I know how to start.
- [x] Yes, I have the time but I will need guidance.
- [ ] No, I don't have the time, but my company or I are [supporting Sequelize through donations on OpenCollective](https://opencollective.com/sequelize).
- [ ] No, I don't have the time, and I understand that I will need to wait until someone from the community or maintainers is interested in resolving my issue.

---

*Indicate your interest in the resolution of this issue by adding the 👍 reaction. Comments such as "+1" will be removed.*

**Comment by papandreou** [comment]
CC @ephys 🤗

**Comment by WikiRik** [comment]
I'm closing this issue already since the PR has been merged. Likely in the next week we'll release a new alpha will includes this fix

**Comment by ucefkh** [comment]
Any news on this?

**Comment by WikiRik** [comment]
The fix landed in alpha 29 so if you download the latest alpha this issue should be resolved

**Comment by ucefkh** [comment]
okay thanks, you mean this version [7.0.0-alpha.9](https://www.npmjs.com/package/sequelize/v/7.0.0-alpha.9)?

thanks

**Comment by WikiRik** [comment]
No, we're switching to `@sequelize/core` as package name in v7; https://www.npmjs.com/package/@sequelize/core?activeTab=readme

**Comment by ucefkh** [comment]
Can't we have this patch please in v7?

**Comment by papandreou** [comment]
I don't think this bug ever existed in v6 -- at least I never experienced it there. I think it was introduced in #15598, released in v7.0.0-alpha.24 and fixed in v7.0.0-alpha.29. Have you actually experienced it with 6.x? If you have, I think you should open a new issue with a reproduction.
